gateway-service-impl/src/main/java/org/hypertrace/gateway/service/common/util/AttributeMetadataUtil.java

```diff
@@ -128,9 +119,32 @@ public static Map<String, AttributeMetadata> remapAttributeMetadataByResultKey(
             expression ->
                 Map.entry(
                     ExpressionReader.getSelectionResultName(expression).orElseThrow(),
-                    attributeMetadataByIdMap.get(
-                        ExpressionReader.getAttributeIdFromAttributeSelection(expression)
-                            .orElseThrow())))
+                    getAttributeMetadata(expression, attributeMetadataByIdMap).orElseThrow()))
         .collect(Collectors.toUnmodifiableMap(Entry::getKey, Entry::getValue, (x, y) -> x));
   }
+
+  private static Optional<AttributeMetadata> getAttributeMetadata(
+      Expression expression, Map<String, AttributeMetadata> attributeMetadataByIdMap) {
+    Optional<AttributeMetadata> maybeAttributeMetadata =
+        ExpressionReader.getAttributeIdFromAttributeSelection(expression)
+            .filter(attributeMetadataByIdMap::containsKey)
+            .map(attributeMetadataByIdMap::get);
+
+    if (maybeAttributeMetadata.isEmpty()) {
+      return Optional.empty();
+    }
+
+    AttributeMetadata attributeMetadata = maybeAttributeMetadata.get();
+    if (!ExpressionReader.isSimpleAttributeSelection(expression)) {
+      switch (attributeMetadata.getValueKind()) {
+        case TYPE_STRING_MAP:
+          return Optional.of(
+              AttributeMetadata.newBuilder(attributeMetadata)
```

Comment:
I don't love creating a fake attribute metadata just to hold kind, but I do like this solution better in general. What was the issue you ran in when working with kinds rather than the general metadata object?

Reply:
The issue was mostly the change size. It was leading to lots of refactor, and tests. Given, that it's a P1, I came up with a short workaround right now for this by generating fake metadata (_since we already follow that approach somewhere else_)

I can take the refactor up as a part of separate PR